server/commands/write.go

```diff
@@ -60,9 +61,43 @@ func (c *WriteCommand) validateTuplesets(ctx context.Context, req *openfgapb.Wri
 	}
 
 	for _, tk := range writes {
-		if _, err := tupleUtils.ValidateTuple(ctx, c.datastore, store, modelID, tk, dbCallsCounter); err != nil {
+		tupleUserset, err := tupleUtils.ValidateTuple(ctx, c.datastore, store, modelID, tk, dbCallsCounter)
+		if err != nil {
 			return serverErrors.HandleTupleValidateError(err)
 		}
+		// TODO: delete the debugging statement
+		c.logger.Info(fmt.Sprintf("%+v\n", tupleUserset))
+		switch usType := tupleUserset.Userset.(type) {
+		case *openfgapb.Userset_This:
+			continue // no need to check on Direct Relationship
+		case *openfgapb.Userset_Intersection:
+			err := isDirectIntersection(usType, tk)
+			if err != nil {
+				return serverErrors.HandleTupleValidateError(err)
+			}
+			continue
+		case *openfgapb.Userset_Union:
+			err := isDirectUnion(usType, tk)
+			if err != nil {
+				return serverErrors.HandleTupleValidateError(err)
+			}
+			continue
+		case *openfgapb.Userset_Difference:
+			err := isDirectDifference(usType, tk)
+			if err != nil {
+				return serverErrors.HandleTupleValidateError(err)
+			}
+			continue
+		case *openfgapb.Userset_ComputedUserset:
+			// if Userset.type is a ComputedUserset then we know it can't be direct
+			err := &tupleUtils.IndirectWriteError{Reason: "Attempting to write directly to an indirect only relationship", TupleKey: tk}
```

Comment:
I think what the linter is saying here is correct. You can just return the error immediately in this case.

Reply:
Yep, totally agree -- the `case... return {}` signature now matches throughout this new section